pkg/planner/core/plan_cache_instance.go

```diff
@@ -0,0 +1,220 @@
+		}
+		if matchCachedPlan(sctx, node.value, matchOpts) { // v.Plan is read-only, no need to lock
+			node.lastUsed.Store(time.Now()) // atomically update the lastUsed field
+			return node.value, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the key and values into the cache.
+// Due to some thread-safety issues, this Put operation might fail, use the returned succ to indicate it.
+func (pc *instancePlanCache) Put(sctx sessionctx.Context, key string, value, opts any) (succ bool) {
+	vMem := value.(*PlanCacheValue).MemoryUsage()
+	if vMem+pc.totCost.Load() > pc.hardMemLimit.Load() {
+		return // do nothing if it exceeds the hard limit
+	}
+	headNode := pc.getHead(key, true)
+	if _, ok := pc.getPlanFromList(sctx, headNode, opts); ok {
+		return // some other thread has inserted the same plan before
+	}
+
+	firstNode := headNode.next.Load()
+	currNode := pc.createNode(value)
+	currNode.next.Store(firstNode)
+	if headNode.next.CompareAndSwap(firstNode, currNode) { // if failed, some other thread has updated this node,
+		pc.totCost.Add(vMem) // then skip this Put and wait for the next time.
+		succ = true
+	}
+	return
+}
+
+// Evict evicts some values. There should be a background thread to perform the eviction.
+// step 1: iterate all values to collect their last_used
+// step 2: estimate an eviction threshold time based on all last_used values
+// step 3: iterate all values again and evict qualified values
+func (pc *instancePlanCache) Evict(_ sessionctx.Context) (evicted bool) {
+	pc.evictMutex.Lock() // make sure only one thread to trigger eviction for safety
+	defer pc.evictMutex.Unlock()
+	if pc.totCost.Load() < pc.softMemLimit.Load() {
+		return // do nothing
+	}
+	lastUsedTimes := make([]time.Time, 0, 64)
+	pc.foreach(func(_, this *instancePCNode) bool { // step 1
+		lastUsedTimes = append(lastUsedTimes, this.lastUsed.Load())
+		return false
+	})
+	threshold := pc.calcEvictionThreshold(lastUsedTimes) // step 2
+	pc.foreach(func(prev, this *instancePCNode) bool {   // step 3
+		if !this.lastUsed.Load().After(threshold) { // if lastUsed<=threshold, evict this value
+			if prev.next.CompareAndSwap(this, this.next.Load()) { // have to use CAS since
+				pc.totCost.Sub(this.value.MemoryUsage()) //  it might have been updated by other thread
+				evicted = true
+				return true
+			}
+		}
+		return false
+	})
+
+	// post operation: clear empty heads in pc.Heads
+	keys, headNodes := pc.headNodes()
+	for i, headNode := range headNodes {
+		if headNode.next.Load() == nil {
+			pc.heads.Delete(keys[i])
+		}
+	}
+	return
+}
+
+// MemUsage returns the memory usage of this plan cache.
+func (pc *instancePlanCache) MemUsage(_ sessionctx.Context) int64 {
+	return pc.totCost.Load()
+}
+
+func (pc *instancePlanCache) calcEvictionThreshold(lastUsedTimes []time.Time) (t time.Time) {
+	avgPerPlan := pc.totCost.Load() / int64(len(lastUsedTimes))
+	if avgPerPlan <= 0 {
+		return
+	}
+	memToRelease := pc.totCost.Load() - pc.softMemLimit.Load()
```

Comment:
Since `pc.totCost` could change, I think we should `Load()` its value only once.